### DISCUSSION TITLE
Création Son sur présence alarme

### DISCUSSION THREAD

**Discussion by MatthM2I** [witsa/synapps#721] [Q&A]
Bonjour,

Je souhaites créer un son sur l'interface a l'apparition d'une alarme dans le journal. Ce son serait désactivable grâce à un bouton d'acquittement sur l'interface. Cela afin d'alerter des télésurveilleurs surveillants plusieurs GTC en même temps. 

Est ce que cela est possible?

Merci de votre retour,
Cordialement,

**Comment by manufitoussiwit** [comment]
Bonjour,

En ce qui concerne le son, je peux vous proposer un composite  : 

```text
SYNAPPS-STUDIO-COMPOSITE|{"config":{"key":"sound","name":"Son","properties":{"height":"0px","width":"0px"},"additionalDefs":{"sound":{"type":"text","label":"Son","helperTxt":"URL du son. Pour générer un dataURI : <a href='https://www.adminbooster.com/tool/data_uri' target='_blank'>File to data URI converter</a>","value":""},"isAutoPlay":{"type":"boolean","label":"Auto play?","value":false},"isLoop":{"type":"boolean","label":"En boucle?","value":false},"delay":{"type":"number","label":"Délai entre boucle","value":500,"step":10,"unit":"ms"}}},"leadActor":{"type":"layout/stack","key":"sound-root","additionalDefs":{"sound":{"type":"text","label":"Son"},"isAutoPlay":{"type":"boolean","label":"Auto play"},"isLoop":{"type":"boolean","label":"En boucle?"}},"bindings":{"additionals.sound":"stage@properties.sound","additionals.isAutoPlay":"stage@properties.isAutoPlay","additionals.isLoop":"stage@properties.isLoop"},"events":{"onInit":["const _ = {","  sound: null,","  isPlaying: false,","  setup: () => {","    if (_.sound) {","      _.sound.removeEventListener('play', _.onPlay);","      _.sound.removeEventListener('ended', _.onEnd);","    }","    if (this.additionals.sound) {","      _.sound = new Audio(this.additionals.sound);","      if (this.additionals.isLoop) {","        _.sound.addEventListener('ended', _.onEnd);","      }","","      if (this.additionals.isAutoPlay) {","        _.play();","      }","    }","  },","","","  destroy() {","    if (_.sound) {","      _.sound.removeEventListener('play', _.onPlay);","      _.sound.removeEventListener('ended', _.onEnd);","      _.sound.pause();","      _.sound = null;","    }","","  },","","  onPlay() {","    _.isPlaying = true;","  },","","  onEnd: async () => {","    _.isPlaying = false;","    if (this.additionals.isLoop) {","      await context.utils.wait(this.stage.properties.delay);","      _.sound.play();","    }","","  },","","  play: async () => {","    _.isPlaying = false;","    _.sound.play();","  },","","  pause() {","    _.sound.pause();","  }","};","","this.stage._ = _;"],"onPostInit":["this.stage._.setup();"],"onPropertyChanged":["if ([","  'additionals.sound',","  'additionals.isAutoPlay',","  'additionals.isLoop',","].includes(context.propertyPath)) {","  this.stage._.setup();","}"],"onDestroy":["this.stage._.destroy();"]}}}
```
Il permet de lui renseigner un son en data URI, de lancer le son automatiquement, de boucler, de mettre un délai entre les boucles.

![image](https://github.com/witsa/synapps/assets/35595723/450ff8c2-0cbe-4fff-9a41-51b999645f94)


Pour le son en data URI, je conseillerais d'utiliser ce site : https://www.adminbooster.com/tool/data_uri 

On fait glisser un son (mp3 ou wav) et ça génère un texte : c'est le data URI à copier/coller dans le champ du composite.

Par script ou liaison, on peut agir sur autoplay du composite mais le mieux est de déclencher par script : 

```js

// si le composite s'appelle sound1 :
const sound = context.getActor('sound1');

// pour le démarrer
sound._.play();

// pour l'arrêter
sound._.pause();

// Lorsque le son tourne, le champs
// sound._.isPlaying est true.

```

>⚠️ il y a une limitation : le son ne démarrera que si l'utilisateur à eu une interaction avec la page. C'est une sécurité des navigateurs. Une interaction, c'est un clic ou autre chose dans le genre.

**Comment by manufitoussiwit** [comment]
En ce qui concerne le système avec le journal, j'ai fait un essai qui pourrait vous inspirer : 

Scène à télécharger : https://1drv.ms/t/s!Ar07ZtaKH5h9iUDiW7iKbZeW2e_3?e=JlkBlG

![image](https://github.com/witsa/synapps/assets/35595723/96bdb3b5-a5bc-4489-9d36-b6ec42fd434a)

L'idée ici est de prendre dans un acteur journal que les évènements en-cours non acquittés. Un script surveille leur nombre. S'il y a en au moins 1, le son se déclenche. Sinon il s'arrête. 

![image](https://github.com/witsa/synapps/assets/35595723/56a9c26d-be07-46bc-af74-be427f312ad7)


Il y a un bouton qui lorsqu'on clique dessus, acquitte le premier évènement du journal et rafraichit le journal.

![image](https://github.com/witsa/synapps/assets/35595723/797ba469-6f0b-4962-b3ce-5e32dda20894)

### 🚨 Attention c'est un exemple
**je ne peux pas vous assurer que ce sera d'une fiabilité parfaite**, notamment à cause de la limitation du navigateur mais ça peut vous donner des idées.

**Comment by MatthM2I** [comment]
Bonjour,

Merci beaucoup du retour rapide!
Je n'ai pas rencontré de dysfonctionnement pour le moment, à voir à l'usage.

Par contre, au niveau de ce script:
![image](https://github.com/witsa/synapps/assets/146428448/688eb8ed-3c4d-41e3-9e56-b0331fcce5f9)
Comment enclencher le son sur apparition d'une nouvelle alarme?  Plutôt que sur présence d'une alarme non acquittée

Pour l'arrêter, j'ai changé l'action du bouton en stoppant directement "sound1", cela fonctionne, mais à chaque rechargement du journal portant l'évènement d'enclenchement du son, celui-ci revient (car présence de l'alarme non acquittée)

L'idée, c'est que le télésurveilleur reçoive un son pour l'interpeler, il va sur l'interface qui tourne en fond, il identifie l'alarme, il prévient l'exploitant, puis il coupe le son.
Ensuite l'exploitant acquitte l'alarme pour prise en compte de l'ordre (sans aucune incidence sur le son)
Si une autre alarme apparait alors que la première n'a pas encore été acquittée, il faudrait que le télésurveilleur soi de nouveau alerté par le son

J'espère être clair :)

**Reply by JeremieDENAEGEL** [reply]
Bonjour,
Pour le déclenchement et l'acquit de l'alarme sonore, cela pourrait être plus simple avec une fonction bascule RS dans une ressource. La sortie .output de la ressource pourrait être le déclencheur du composite de @manufitoussiwit

Pour un site similaire au tien, les utilisateurs de mon site ont un acquit visuel à faire en cas de nouvelles alarmes dans le synapps et cet acquittement est enregistré dans le journal avec le nom de l'utilisateur qui a appuyé.
Je te mets ci-dessous mon bouton synapps et les captures d'écran de mes 2 ressources. 
A voir suivant ta programmation si cette méthode est possible ;)

````
SYNAPPS-STUDIO-ACTOR|{"type":"input/button","key":"BP-Acquit","properties":{"backgroundColor":"white","height":"5vmin","width":"5vmin","verticalAlignment":"middle","backgroundPosition":"center center","backgroundSize":"contain","backgroundRepeat":"no-repeat","borderColor":"black","borderWidth":"0.1vmin","borderStyle":"solid","borderRadius":"1vmin","toolTip":"Cliquez pour acquitter les alarmes"},"additionalDefs":{"BP":{"type":"boolean","label":"Lien Ress Cons.Logique Acquittement"},"delay":{"type":"number","label":"Délai impulsion","min":1000,"step":100,"unit":"ms"},"ETAT":{"type":"boolean","label":"Etat Ress Synthèse Détection alarme"},"ETAT1":{"type":"picture","label":"ETAT = 1","helperTxt":"image si Etat Ress Synthèse Détection alarme = 1"},"ETAT0":{"type":"picture","label":"ETAT = 0","helperTxt":"image si Etat Ress Synthèse Détection alarme = 0"}},"additionals":{"delay":2000},"bindings":{"additionals.BP":{"sourceType":"redy/wos-relative-variable","canWrite":true,"relativeTo":"global/resources","relativePath":"WOP.AcquitAlarme.SetPoint","fieldName":"value","dataReadMode":"always","writeOnChange":true},"additionals.ETAT":{"sourceType":"redy/wos-relative-variable","relativeTo":"global/resources","relativePath":"WOP.R01000.Output","fieldName":"value","dataReadMode":"always"},"additionals.ETAT1":{"sourceType":"picture","path":"BELL5"},"additionals.ETAT0":{"sourceType":"picture","path":"BELL2"}},"events":{"onClick":["this.additionals.BP = 1"],"onPropertyChanged":["if (this.additionals.ETAT == 1){","  this.properties.backgroundImage = this.additionals.ETAT1","}else{","  this.properties.backgroundImage = this.additionals.ETAT0","}"]}}
````
![DetectionAlarmeEnCours](https://github.com/witsa/synapps/assets/50445069/009ac2b8-c8d9-4b7c-bbcb-574a008010d6)
![Conslogique](https://github.com/witsa/synapps/assets/50445069/a622fb00-2e84-421d-98c5-d09bff6918fa)

**Comment by MatthM2I** [comment]
Bonjour,

Merci pour ton aide

Effectivement, passer par une ressource est possible également
Mais je n'arrive pas à trouver la solution d'enclenchement à l'apparition d'une nouvelle alarme
Admettons une alarme est déjà présente, non acquittée et le son désactivé -> jusque là c'est facile à mettre en œuvre
Mais quand une nouvelle alarme apparait, comment relancer une nouvelle alerte sonore? Et via ressource ou Synapps?

**Comment by JeremieDENAEGEL** [comment]
Bonjour,
Les éléments que je t'ai donné répondent à ta demande. Après cela dépend aussi de la manière dont tu as réalisé ta programmation.
Une nouvelle alarme peut parfaitement déclencher à nouveau le SET de la bascule RS et donc le son

**Comment by MatthM2I** [comment]
Bonjour,

C'est que ta variable "détection Alarme" sur ton schéma est donc une impulsion?
Si oui, d'où vient elle?

**Reply by JeremieDENAEGEL** [reply]
On rentre là dans de la programmation et non de l'imagerie.
Cependant pour le cas que tu as évoqué, je t'invite à regarder du côté des zones, cela devrait répondre à ton besoin.

**Reply by MatthM2I** [reply]
OK je regarde de ce coté, merci des infos
